### ISSUE TITLE
Login procedure redundantly asks to login by email before logging in with GitHub

### ISSUE THREAD

**Issue by Hexstream** [ultralisp/ultralisp#27]
1. Go to https://ultralisp.org/
2. Click "Add projects from GitHub".
Page saying "Please login to add GitHub repositories to Ultralisp." appears.
3. Click "Log in".

At this point (or maybe even right after clicking "Add projects from GitHub"), I would expect to log in with GitHub, but the page instead says "Enter email and we'll send a link with a code to login:" and you have to enter your email address and go click the link in the email, and only after that are you allowed to login with GitHub.

I think this workflow should get a little refactoring so that you can directly login with GitHub without going through the email steps if you want...

**Comment by svetlyak40wt** [comment]
I thought about this too. Actually, here user does not login using GitHub, it authorizes the app to do something with GitHub.

When I started this project, I considered to add a "login with GitHub" option, but:

* there could be a situation when a user doesn't have an account because he uses Bitbucket or Gitlab;
* GitHub not always give an email for an authenticated user, but Ultralisp needs a user's email to send him notification about broken builds. And if I did log in with GitHub, then I have to implement additional flow to collect an email from the user.
* If login with GitHub was implemented, then I have to implement a bunch of other "login with Google", "login with Bitbucket", "login with Facebook", etc. But I don't want to.

That is why I decided to make a single way to log in – via a link sent by email.

Probably, we need to reformulate some text on a page to make a distinction between logging in and authorization access to the GitHub.

**Comment by Hexstream** [comment]
> * there could be a situation when a user doesn't have an account because he uses Bitbucket or Gitlab

I bet most people who use Bitbucket or GitLab also have a GitHub account anyway, and Ultralisp is already fairly GitHub-centric. We lose nothing by streamlining the GitHub flow further...

> * GitHub not always give an email for an authenticated user, but Ultralisp needs a user's email to send him notification about broken builds. And if I did log in with GitHub, then I have to implement additional flow to collect an email from the user.

I would make a user settings page that uses the email provided by GitHub by default if available, and lets the user choose another email address if they wish.

> * If login with GitHub was implemented, then I have to implement a bunch of other "login with Google", "login with Bitbucket", "login with Facebook", etc.

... No? Why? "Login with Google" and "Login with Bitbucket" would be nice-to-haves, "Login with Facebook" would probably be worse than worthless. I think it's fine to be a bit GitHub-centric at first, and that's already the case...

**Comment by ryukinix** [comment]
Just to report, I can replicate the same issue here.

**Comment by svetlyak40wt** [comment]
Actually, as I wrote above. Right now this is a feature. But I'm ready to replace login via email by login by GitHub in near future.

**Comment by svetlyak40wt** [comment]
I think, this is the next feature I'll work on.

**Comment by ryukinix** [comment]
Great to know! This is a nice feature.

**Comment by Hexstream** [comment]
Great choice, this project is going at *light speed!!* :heart:

**Comment by svetlyak40wt** [comment]
Guys, I've done. But this issue wasn't "good first issue" and required a lot of work.

As a result, I've created a new authentication library for Weblocks: [weblocks-auth](https://github.com/40ants/weblocks-auth).

I'm closing the issue because all changes are in production now.

**Comment by Hexstream** [comment]
Great! I tried it just now, this is a HUGE increase in usability!
This was especially important to fix since it greatly affects the onboarding experience.

The "[My projects](https://ultralisp.org/my/projects)" link when logged in leads to a 404, I'm assuming that's a not yet implemented feature? If so, maybe put an actual placeholder page there...

I noticed that the "Write repository hooks" permission is currently required to login.
That may be slightly concerning to some, so maybe it should be made optional?
That would be one additional way to avoid scaring the user during onboarding.

Also consider removing the "good first issue" label. :)

**Comment by svetlyak40wt** [comment]
> I noticed that the "Write repository hooks" permission is currently required to login.
That may be slightly concerning to some, so maybe it should be made optional?
That would be one additional way to avoid scaring the user during onboarding.

Github shows you all scopes which ever have been granted. For new users, only permission to read their email will be asked on first login.

You can check this yourself. Go to https://github.com/settings/applications page, select Ultralisp and revoke permissions. Then try to log in again. You should see only one requested permission:

<img width="585" alt="Screen Shot 2019-04-19 at 21 31 24" src="https://user-images.githubusercontent.com/24827/56438172-aa448f80-62ea-11e9-8904-42554f7a72eb.png">

**Comment by svetlyak40wt** [comment]
Broken link from "My projects" is the result of my work on this issue:

https://github.com/ultralisp/ultralisp/issues/31

I've added a link but forgot to add a page.

This should be fixed: https://github.com/ultralisp/ultralisp/issues/44

**Comment by Hexstream** [comment]
Sorry, [you are correct](https://github.com/ultralisp/ultralisp/issues/27#issuecomment-484980485). I got confused. I apologize.
